### ISSUE TITLE
Passing the r function itself in rinterface.rternalize through multiprocessing.Queue crashes the queue

### ISSUE THREAD

**Issue by DE0CH** [rpy2/rpy2#970]
**Describe the issue or bug**

When passing a function decorated with `rinterface.rternalize` through `multiprocessing.Queue` in the function itself, the queue just becomes empty forever without any error raised.

**To Reproduce**

```
from rpy2.robjects.packages import importr # even though importr is not used, it is here to avoid triggering #971 bug
import rpy2.rinterface as ri
from multiprocessing import Queue

def make_function(q, s):
    @ri.rternalize
    def putter():
        q.put(s)
        q.get()
        # It gets stuck here
        print("passed")
    return putter

if __name__ == '__main__':
    q = Queue()
    s = [None]
    putter = make_function(q, s)
    s[0] = putter # This line is the culprit, if we comment out this line, everything works as expected.
    putter()
```

**Expected behavior**

I expect the Queue to be able to pass it just like any other object, if it cannot for some reason, it should raise an exception.

**Error**

There is no error. It just gets stuck.

**Additional context**
```
rpy2 version:
3.5.6
Python version:
3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110]
Looking for R's HOME:
    Environment variable R_HOME: None
    Calling `R RHOME`: /usr/lib/R
    Environment variable R_LIBS_USER: None
R's additions to LD_LIBRARY_PATH:
/usr/lib/R/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/default-java/lib/server
R version:
    In the PATH: R version 4.0.4 (2021-02-15) -- "Lost Library Book"
    Loading R library from rpy2: OK
Additional directories to load R packages from:
None
C extension compilation:
  include:
  ['/usr/share/R/include']
  libraries:
  ['R', 'pcre2-8', 'lzma', 'bz2', 'z', 'rt', 'dl', 'm', 'icuuc', 'icui18n']
  library_dirs:
  ['/usr/lib/R/lib']
  extra_compile_args:
  []
  extra_link_args:
  ['-Wl,--export-dynamic', '-fopenmp', '-Wl,-z,relro']
```

**Comment by EthanKesslerDev** [comment]
What causes no segmentation fault to be raised, and instead just creates the empty queue?

**Comment by DE0CH** [comment]
I'm not really sure. I've seen it in a very complex usage. However, I don't think it's that important to reproduce it minimally because fixing m the segfault will most likely fix it.

**Comment by DE0CH** [comment]
Actually, when I was trying to reproduce this empty queue, I discovered that the segfault is actually not caused by the queue but `rpy2.rinterface.rternalize`. So filed #971 instead. Going to close this for now and reopen it after I figure out how I can reproduce it.

**Comment by DE0CH** [comment]
Just edited the main issue with the correct reproducible example.

**Comment by lgautier** [comment]
Putting R functions that have been "rternalized" can only work if the serialization mechanism used by Python's `multiprocessing.Queue` is also able to move objects across processes. `rpy2` make R object "pickle-able" (by wrapping R's own object serialization system) so this should be all that is needed (see https://bugs.python.org/issue28973).

This can be confirmed with the following modified example
```python
import rpy2.rinterface as ri
ri.initr()  # See 
from multiprocessing import Queue

def make_function(q, s):
    @ri.rternalize
    def putter():
        q.put(s)
        q.get()
        print("passed")
    return putter

if __name__ == '__main__':
    q = Queue()
    s = [None]
    putter = make_function(q, s)
    
    # Assert that the putter object is pickleable
    import tempfile
    import pickle
    with tempfile.TemporaryFile() as fh:
        pickle.dump(putter, fh)
        
    
```

So what is happening has likely to do with assumptions `Queue` is making; either it is not used the way it is intended to or your use-case uncover a flaw it is design. Thinking about the overall flow across Python and R can likely get you to a simpler design that either can work or help troubleshoot. As it is I  don't understand what you are trying to do.

One way to go about it is to think about R as just a Python library to delegate compute tasks to.  Create the following function if this helps:

```python
def what_will_happen_in_r(*args, **kwargs):
    print('Compute something in R.')
```

With that it mind you can first start to write your multiprocessing code *without* rpy2 and first get it to work.

**Comment by DE0CH** [comment]
I see now. Also adding the line `s[0] = putter` is also pickable. 

My use case is quite weird, I'm writing a library that basically takes a user defined function (called `targetRunner`) in `senario` and the package calls it with some parameters as well as the `senario`. Because of this bug, we just remove it before giving it back https://github.com/auto-optimization/iracepy/blob/9eb2923f63344e9ab05bb3625318c79ea372dd7c/src/irace/__init__.py#L157-L159

But it's weird that when the queue crashes, it doesn't throw any error, but just gets stuck. Do you think this is a bug in `multiprocess.Queue` and not `rpy2`?

**Comment by lgautier** [comment]
> I see now. Also adding the line `s[0] = putter` is also pickable.
> 
> My use case is quite weird, I'm writing a library that basically takes a user defined function (called `targetRunner`) in `senario` and the package calls it with some parameters as well as the `senario`. Because of this bug, we just remove it before giving it back https://github.com/auto-optimization/iracepy/blob/9eb2923f63344e9ab05bb3625318c79ea372dd7c/src/irace/__init__.py#L157-L159

I am not completely following the idea of what `targetRunner` or `scenario` represent. Thinking about the code path (call stack) across Python and R will help isolate the issue, and possibly indicate whether this is more complicated than it has to be. In the sequence Python(i) -> R -> Python(ii) the latest step has an rternalized function that is defined with a closure. Closures and multiprocessing are not handled well by default in Python:
 
```python
from multiprocessing import Queue

def make_function(q, s):
    def putter():
        q.put(s)
        q.get()
        print("passed")
    return putter

if __name__ == '__main__':
    q = Queue()
    s = [None]
    putter = make_function(q, s)
    s[0] = putter
    putter()
```

It hangs indefinitely and prints an error:
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 239, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
AttributeError: Can't pickle local object 'make_function.<locals>.putter'
```

With that in mind, you can try to simplify and try without a closure to see what happens:

```python
import rpy2.rinterface as ri
ri.initr()  # See 
from multiprocessing import Queue

@ri.rternalize
def putter():
    q.put(s)
    #  q.get()  # hangs when not commented out.
     
if __name__ == '__main__':
    q = Queue()
    s = [None]
    s[0] = putter
    putter()
```

The code above completes without error but prints the following:
```
ValueError: rternalized None
```

This is caused by an exception raised during the evaluation of the "rternalized" function: https://github.com/rpy2/rpy2/blob/master/rpy2/rinterface_lib/_rinterface_capi.py#L621. Currently rpy2 is logging the exception rather than propagating an exception back across the call stack across languages. This caused by `rpy2` being unsure about how to convert a Python `None` into.  That message can be suppressed by returning an object `rpy2.rinterface` knows how to convert.

```python
@ri.rternalize
def putter():
    q.put(s)
    #  q.get()  # hangs when not commented out.
    return 0
```

With that simple form and no error report printed, I get the following when interrupting when it hangs with `q.get()`:
 
```
  File "test_issue970.py", line 8, in putter
    # q.get()  # hangs when not commented out.
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 97, in get
    res = self._recv_bytes()
  File "/usr/lib/python3.8/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/usr/lib/python3.8/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/usr/lib/python3.8/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
  File "/home/laurent/Desktop/software/python/rpy2_github/rpy2/rinterface.py", line 94, in _sigint_handler
    raise KeyboardInterrupt()
KeyboardInterrupt: 
```

After a quick search about the above you should converge on `multiprocessing.queues.SimpleQueue` as a simpler alternative when `Queues` appear stuck in a deadlock. A possible reason here is around rpy2's use of `threading.Rlock` when operating with memory management tasks that would cause segfaults that would be both sudden and cause debugging challenges.

The following snippet of code is now completing.

```python
import rpy2.rinterface as ri
ri.initr() 
import multiprocessing
from multiprocessing.queues import SimpleQueue


@ri.rternalize
def putter():
    q.put(s)
    q.get()  # hangs when not commented out.
    return 0


if __name__ == '__main__':
    q = SimpleQueue(ctx=multiprocessing.get_context())
    s = [None]
    s[0] = putter
    putter()
```

Reintroducing the complexity in your original code (rternalized function with a closure) is left as an exercise for the reader.

> But it's weird that when the queue crashes, it doesn't throw any error, but just gets stuck. Do you think this is a bug in `multiprocess.Queue` and not `rpy2`?

I definitely consider the segfault when calling `rternalize` before R is initialized a bug. Even if users are not supposed to do it this should fail with a Python exception raised. This now fixed in rpy2 3.5.7. There is probably an argument to make that Python's `multiprocess.Queue` should not hang but it could also be more difficult to achieve that it first seems. If this is a bug in Python my intuition is that it is a known one and there is an open issue about it.

I am closing this as either brittleness in Python's API for `Queue`, or operator error.
